contracts/token/ERC6909/draft-ERC6909.sol

```diff
@@ -0,0 +1,168 @@
+    }
+
+    /// @inheritdoc IERC6909
+    function approve(address spender, uint256 id, uint256 amount) external virtual override returns (bool) {
+        address caller = _msgSender();
+        _allowances[caller][spender][id] = amount;
+
+        emit Approval(caller, spender, id, amount);
+        return true;
+    }
+
+    /// @inheritdoc IERC6909
+    function setOperator(address spender, bool approved) external virtual override returns (bool) {
+        address caller = _msgSender();
+        _operatorApprovals[caller][spender] = approved;
+
+        emit OperatorSet(caller, spender, approved);
+        return true;
+    }
+
+    /// @inheritdoc IERC6909
+    function transfer(address receiver, uint256 id, uint256 amount) external virtual override returns (bool) {
+        _transfer(_msgSender(), receiver, id, amount);
+        return true;
+    }
+
+    /// @inheritdoc IERC6909
+    function transferFrom(
+        address sender,
+        address receiver,
+        uint256 id,
+        uint256 amount
+    ) external virtual override returns (bool) {
+        address caller = _msgSender();
+        if (caller != sender && !isOperator(sender, caller)) {
+            uint256 currentAllowance = allowance(sender, caller, id);
+            if (currentAllowance != type(uint256).max) {
+                if (currentAllowance < amount) {
+                    revert ERC6909InsufficientAllowance(caller, currentAllowance, amount, id);
+                }
+                unchecked {
+                    _allowances[sender][caller][id] = currentAllowance - amount;
+                }
+            }
+        }
+
+        _transfer(sender, receiver, id, amount);
+        return true;
+    }
+
+    /**
+     * @dev Moves `amount` of token `id` from `from` to `to` without checking for approvals.
+     *
+     * This internal function is equivalent to {transfer}, and can be used to
+     * e.g. implement automatic token fees, slashing mechanisms, etc.
+     *
+     * Emits a {Transfer} event.
+     *
+     * NOTE: This function is not virtual, {_update} should be overridden instead.
+     */
+    function _transfer(address from, address to, uint256 id, uint256 amount) internal {
+        if (from == address(0)) {
+            revert ERC6909InvalidSender(address(0));
+        }
+        if (to == address(0)) {
+            revert ERC6909InvalidReceiver(address(0));
+        }
+        _update(from, to, id, amount);
+    }
+
+    /**
+     * @dev Transfers `amount` of token `id` from `from` to `to`, or alternatively mints (or burns) if `from`
+     * (or `to`) is the zero address. All customizations to transfers, mints, and burns should be done by overriding
+     * this function.
+     *
+     * Emits a {Transfer} event.
+     */
+    function _update(address from, address to, uint256 id, uint256 amount) internal virtual {
+        address caller = _msgSender();
+
+        if (from != address(0)) {
+            uint256 fromBalance = balanceOf(from, id);
```

Comment:
IMO its important to keep the internal account separate from the exposed values. In most cases it doesn't change anything, but if someone ever wants to do some rebasing, you don't want to have to deal with that part. 
```suggestion
            uint256 fromBalance = _balances[id][from];
```

See ERC20.sol and ERC1155.sol